From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ken Russell <kbr@chromium.org>
Date: Fri, 28 Aug 2026 19:33:51 -0700
Subject: [M153] GL: add finish-before-multi-attachment-blit workaround.

Original change's description:
> GL: add finish-before-multi-attachment-blit workaround.
>
> If performing glBlitFramebuffer to a draw framebuffer which either has
> multiple color attachments, or has a color texture attachment with
> base level greater than zero, glFinish() as a workaround. Apply this
> to Mali GPUs.
>
> It wasn't feasible to add a regression test for this workaround.
>
> Bug: chromium:546252753
> Change-Id: I3ef79ac6f0d25529bd22db598262704d2056802b
> Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/8311814
> Commit-Queue: Geoff Lang <geofflang@chromium.org>
> Auto-Submit: Kenneth Russell <kbr@chromium.org>
> Reviewed-by: Zhenyao Mo <zmo@chromium.org>
> Reviewed-by: Geoff Lang <geofflang@chromium.org>

(cherry picked from commit 8efd15f71c27cd0bc2a9cf0074d77e899ca9c448)

Bug: chromium:555567586,chromium:546252753
Change-Id: I3ef79ac6f0d25529bd22db598262704d2056802b
Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/8341207
Reviewed-by: Geoff Lang <geofflang@chromium.org>
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>

diff --git a/include/platform/autogen/FeaturesGL_autogen.h b/include/platform/autogen/FeaturesGL_autogen.h
index bd7bb2cdf841810aa7b36c162523e14435b399ae..414d3f02081f659ad383490c23e58841662b4ad4 100644
--- a/include/platform/autogen/FeaturesGL_autogen.h
+++ b/include/platform/autogen/FeaturesGL_autogen.h
@@ -764,6 +764,12 @@ struct FeaturesGL : FeatureSetBase
         &members,
     };
 
+    FeatureInfo finishBeforeBlitFramebufferMultiAttachment = {
+        "finishBeforeBlitFramebufferMultiAttachment",
+        FeatureCategory::OpenGLWorkarounds,
+        &members,
+    };
+
 };
 
 inline FeaturesGL::FeaturesGL()  = default;
diff --git a/include/platform/gl_features.json b/include/platform/gl_features.json
index 96082d8f4a63c508b608eb9757a14806503e41f2..21a76917baaef35f14d37c4fca146e87fad11abf 100644
--- a/include/platform/gl_features.json
+++ b/include/platform/gl_features.json
@@ -1001,6 +1001,15 @@
                 "Reattach texture layer to framebuffer after calling TexImage3D to increase layer count on PowerVR Android."
             ],
             "issue": "http://crbug.com/499602793"
+        },
+        {
+            "name": "finish_before_blit_framebuffer_multi_attachment",
+            "category": "Workarounds",
+            "description": [
+                "Issue a glFinish before calling glBlitFramebuffer if the draw framebuffer has ",
+                "multiple color attachments or any color attachment's texture has base_level > 0."
+            ],
+            "issue": "http://crbug.com/546252753"
         }
     ]
 }
diff --git a/src/libANGLE/renderer/gl/FramebufferGL.cpp b/src/libANGLE/renderer/gl/FramebufferGL.cpp
index 625eefcbb13cf75aa8a9d7f2861e4c57a520e216..ca7caeeab233ae8ade0cf44df4b152faa51747f2 100644
--- a/src/libANGLE/renderer/gl/FramebufferGL.cpp
+++ b/src/libANGLE/renderer/gl/FramebufferGL.cpp
@@ -14,6 +14,7 @@
 #include "libANGLE/ErrorStrings.h"
 #include "libANGLE/FramebufferAttachment.h"
 #include "libANGLE/State.h"
+#include "libANGLE/Texture.h"
 #include "libANGLE/angletypes.h"
 #include "libANGLE/formatutils.h"
 #include "libANGLE/queryconversions.h"
@@ -975,6 +976,36 @@ angle::Result FramebufferGL::blit(const gl::Context *context,
         }
     }
 
+    if (features.finishBeforeBlitFramebufferMultiAttachment.enabled)
+    {
+        bool needFinish = false;
+        if (destFramebuffer->getState().getColorAttachmentsMask().count() > 1)
+        {
+            needFinish = true;
+        }
+        else
+        {
+            for (size_t colorIndex : destFramebuffer->getState().getColorAttachmentsMask())
+            {
+                const FramebufferAttachment *attachment =
+                    destFramebuffer->getColorAttachment(colorIndex);
+                if (attachment && attachment->type() == GL_TEXTURE)
+                {
+                    const Texture *texture = attachment->getTexture();
+                    if (texture && texture->getBaseLevel() > 0)
+                    {
+                        needFinish = true;
+                        break;
+                    }
+                }
+            }
+        }
+        if (needFinish)
+        {
+            functions->finish();
+        }
+    }
+
     functions->blitFramebuffer(finalSourceArea.x, finalSourceArea.y, finalSourceArea.x1(),
                                finalSourceArea.y1(), finalDestArea.x, finalDestArea.y,
                                finalDestArea.x1(), finalDestArea.y1(), blitMask, filter);
diff --git a/src/libANGLE/renderer/gl/renderergl_utils.cpp b/src/libANGLE/renderer/gl/renderergl_utils.cpp
index 9e16d2c5ce16aeb2632251896ee3be494a7b81f1..41b72b28c93719040729a803c40c4bcc9a789c6b 100644
--- a/src/libANGLE/renderer/gl/renderergl_utils.cpp
+++ b/src/libANGLE/renderer/gl/renderergl_utils.cpp
@@ -2685,6 +2685,9 @@ void InitializeFeatures(const FunctionsGL *functions, angle::FeaturesGL *feature
     // http://crbug.com/534468209
     ANGLE_FEATURE_CONDITION(features, flushQueriesBeforeDeletingOrUnbindingFbo, isMali);
 
+    // http://crbug.com/546252753
+    ANGLE_FEATURE_CONDITION(features, finishBeforeBlitFramebufferMultiAttachment, isMali);
+
     // https://crbug.com/40264674
     ANGLE_FEATURE_CONDITION(features, disableClipControl, IsMaliG72OrG76OrG51(functions));
 
diff --git a/util/autogen/angle_features_autogen.cpp b/util/autogen/angle_features_autogen.cpp
index a5c51f6a4ae797bdd024d2ab5751853900f5f1a8..aa37dfc806d2081e56d88827956619dd1122e4ff 100644
--- a/util/autogen/angle_features_autogen.cpp
+++ b/util/autogen/angle_features_autogen.cpp
@@ -184,6 +184,7 @@ constexpr PackedEnumMap<Feature, const char *> kFeatureNames = {{
     {Feature::ExposeES32ForTesting, "exposeES32ForTesting"},
     {Feature::ExposeNonConformantExtensionsAndVersions, "exposeNonConformantExtensionsAndVersions"},
     {Feature::ExternallySynchronizePipelineCacheAccess, "externallySynchronizePipelineCacheAccess"},
+    {Feature::FinishBeforeBlitFramebufferMultiAttachment, "finishBeforeBlitFramebufferMultiAttachment"},
     {Feature::FinishDoesNotCauseQueriesToBeAvailable, "finishDoesNotCauseQueriesToBeAvailable"},
     {Feature::FlushAfterEndingTransformFeedback, "flushAfterEndingTransformFeedback"},
     {Feature::FlushAfterStreamVertexData, "flushAfterStreamVertexData"},
diff --git a/util/autogen/angle_features_autogen.h b/util/autogen/angle_features_autogen.h
index c452c86577004c02bf76a9a0a5e134d38b460891..b225ef96c52c7d47f34e802e4b0a896710d84d22 100644
--- a/util/autogen/angle_features_autogen.h
+++ b/util/autogen/angle_features_autogen.h
@@ -184,6 +184,7 @@ enum class Feature
     ExposeES32ForTesting,
     ExposeNonConformantExtensionsAndVersions,
     ExternallySynchronizePipelineCacheAccess,
+    FinishBeforeBlitFramebufferMultiAttachment,
     FinishDoesNotCauseQueriesToBeAvailable,
     FlushAfterEndingTransformFeedback,
     FlushAfterStreamVertexData,
